File: config-service/src/config_service/repositories/application_repository.py
"""Repository for Application entity data access."""

import json
from typing import List, Optional
from datetime import datetime
from ulid import ULID as ULIDGenerator
from config_service.database.connection import db_manager
from config_service.models.application import Application, ApplicationCreate, ApplicationUpdate, ApplicationWithConfigs


class ApplicationRepository:
    """Repository for Application entity operations using raw SQL."""
    
    async def create(self, application_data: ApplicationCreate) -> Application:
        """Create a new application."""
        import logging
        logger = logging.getLogger(__name__)
        
        try:
            app_id = str(ULIDGenerator())  # Convert to string immediately
            logger.info(f"Generated ULID: {app_id}")
            
            now = datetime.now()
            logger.info(f"Current time: {now}")
            
            query = """
            INSERT INTO applications (id, name, comments, created_at, updated_at)
            VALUES (%s, %s, %s, %s, %s)
            RETURNING id, name, comments, created_at, updated_at
            """
            
            params = (app_id, application_data.name, application_data.comments, now, now)
            logger.info(f"Query params: {params}")
            
            # Use execute_returning_query for INSERT with RETURNING
            results = await db_manager.execute_returning_query(query, params)
            logger.info(f"Query results: {results}")
            
            if not results:
                raise RuntimeError("Failed to create application")
            
            row = results[0]
            logger.info(f"Row data: {row}")
            
            app = Application(
                id=row['id'],
                name=row['name'],
                comments=row['comments'],
                created_at=row['created_at'],
                updated_at=row['updated_at']
            )
            logger.info(f"Created Application object: {app}")
            return app
            
        except Exception as e:
            logger.error(f"Error in create method: {e}", exc_info=True)
            raise RuntimeError(f"Failed to create application: {e}")
    
    async def get_by_id(self, app_id: str) -> Optional[Application]:
        """Get application by ID."""
        query = """
        SELECT id, name, comments, created_at, updated_at
        FROM applications
        WHERE id = %s
        """
        
        try:
            results = await db_manager.execute_query(query, (app_id,))
            if not results:
                return None
            
            row = results[0]
            return Application(
                id=row['id'],
                name=row['name'],
                comments=row['comments'],
                created_at=row['created_at'],
                updated_at=row['updated_at']
            )
        except Exception as e:
            raise RuntimeError(f"Failed to get application: {e}")
    
    async def get_by_id_with_configs(self, app_id: str) -> Optional[ApplicationWithConfigs]:
        """Get application by ID including related configuration IDs."""
        query = """
        SELECT 
            a.id, a.name, a.comments, a.created_at, a.updated_at,
            COALESCE(
                json_agg(c.id) FILTER (WHERE c.id IS NOT NULL), 
                '[]'::json
            ) as configuration_ids
        FROM applications a
        LEFT JOIN configurations c ON a.id = c.application_id
        WHERE a.id = %s
        GROUP BY a.id, a.name, a.comments, a.created_at, a.updated_at
        """
        
        try:
            results = await db_manager.execute_query(query, (app_id,))
            if not results:
                return None
            
            row = results[0]
            config_ids = []
            if row['configuration_ids'] and row['configuration_ids'] != [None]:
                config_ids = [cid for cid in row['configuration_ids']]
            
            return ApplicationWithConfigs(
                id=row['id'],
                name=row['name'],
                comments=row['comments'],
                created_at=row['created_at'],
                updated_at=row['updated_at'],
                configuration_ids=config_ids
            )
        except Exception as e:
            raise RuntimeError(f"Failed to get application with configs: {e}")
    
    async def get_all(self) -> List[Application]:
        """Get all applications."""
        query = """
        SELECT id, name, comments, created_at, updated_at
        FROM applications
        ORDER BY name
        """
        
        try:
            results = await db_manager.execute_query(query)
            return [
                Application(
                    id=row['id'],
                    name=row['name'],
                    comments=row['comments'],
                    created_at=row['created_at'],
                    updated_at=row['updated_at']
                )
                for row in results
            ]
        except Exception as e:
            raise RuntimeError(f"Failed to get applications: {e}")
    
    async def update(self, app_id: str, application_data: ApplicationUpdate) -> Optional[Application]:
        """Update an existing application."""
        now = datetime.now()
        
        query = """
        UPDATE applications
        SET name = %s, comments = %s, updated_at = %s
        WHERE id = %s
        RETURNING id, name, comments, created_at, updated_at
        """
        
        params = (application_data.name, application_data.comments, now, app_id)
        
        try:
            results = await db_manager.execute_returning_query(query, params)
            if not results:
                return None
            
            row = results[0]
            return Application(
                id=row['id'],
                name=row['name'],
                comments=row['comments'],
                created_at=row['created_at'],
                updated_at=row['updated_at']
            )
        except Exception as e:
            raise RuntimeError(f"Failed to update application: {e}")
    
    async def delete(self, app_id: str) -> bool:
        """Delete an application."""
        query = "DELETE FROM applications WHERE id = %s"
        
        try:
            affected_rows = await db_manager.execute_command(query, (app_id,))
            return affected_rows > 0
        except Exception as e:
            raise RuntimeError(f"Failed to delete application: {e}")
    
    async def delete_multiple(self, app_ids: List[str]) -> int:
        """Delete multiple applications by IDs."""
        if not app_ids:
            return 0
        
        # Create placeholders for the IN clause
        placeholders = ','.join(['%s'] * len(app_ids))
        query = f"DELETE FROM applications WHERE id IN ({placeholders})"
        
        try:
            affected_rows = await db_manager.execute_command(query, tuple(app_ids))
            return affected_rows
        except Exception as e:
            raise RuntimeError(f"Failed to delete applications: {e}")


# Global repository instance
application_repository = ApplicationRepository()

File: config-service-dotnet/src/ConfigService.Infrastructure/Repositories/ApplicationRepository.cs
using ConfigService.Core.Interfaces;
using ConfigService.Core.Models;
using ConfigService.Core.Services;
using ConfigService.Infrastructure.Data;
using Microsoft.Extensions.Logging;
using System.Data;

namespace ConfigService.Infrastructure.Repositories;

/// <summary>
/// PostgreSQL implementation of the Application repository
/// </summary>
public class ApplicationRepository : IApplicationRepository
{
    private readonly DatabaseContext _context;
    private readonly ILogger<ApplicationRepository> _logger;

    public ApplicationRepository(DatabaseContext context, ILogger<ApplicationRepository> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<Application> CreateAsync(ApplicationCreate applicationData)
    {
        var id = UlidGenerator.NewUlid();
        var now = DateTime.UtcNow;

        _logger.LogInformation("Creating application: name='{Name}' comments='{Comments}'", 
            applicationData.Name, applicationData.Comments);
        _logger.LogInformation("Generated ULID: {Id}", id);
        _logger.LogInformation("Current time: {Time}", now);

        const string sql = @"
            INSERT INTO applications (id, name, comments, created_at, updated_at)
            VALUES (@Id, @Name, @Comments, @CreatedAt, @UpdatedAt)
            RETURNING id, name, comments, created_at, updated_at";

        var parameters = new
        {
            Id = id,
            Name = applicationData.Name,
            Comments = applicationData.Comments,
            CreatedAt = now,
            UpdatedAt = now
        };

        try
        {
            var result = await _context.QuerySingleOrDefaultAsync(sql, parameters, MapApplication);
            
            if (result == null)
            {
                throw new InvalidOperationException("Failed to create application - no result returned");
            }

            _logger.LogInformation("Application created successfully: {Application}", result);
            return result;
        }
        catch (Exception ex) when (ex.Message.Contains("unique constraint") || ex.Message.Contains("duplicate"))
        {
            _logger.LogError(ex, "Unique constraint violation creating application");
            throw new InvalidOperationException($"Application with name '{applicationData.Name}' already exists", ex);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating application");
            throw;
        }
    }

    public async Task<Application?> GetByIdAsync(string id)
    {
        const string sql = "SELECT id, name, comments, created_at, updated_at FROM applications WHERE id = @Id";
        
        return await _context.QuerySingleOrDefaultAsync(sql, new { Id = id }, MapApplication);
    }

    public async Task<ApplicationWithConfigs?> GetByIdWithConfigsAsync(string id)
    {
        const string sql = @"
            SELECT 
                a.id, a.name, a.comments, a.created_at, a.updated_at,
                COALESCE(
                    ARRAY_AGG(c.id ORDER BY c.name) FILTER (WHERE c.id IS NOT NULL), 
                    ARRAY[]::VARCHAR[]
                ) as configuration_ids
            FROM applications a
            LEFT JOIN configurations c ON a.id = c.application_id
            WHERE a.id = @Id
            GROUP BY a.id, a.name, a.comments, a.created_at, a.updated_at";
        
        return await _context.QuerySingleOrDefaultAsync(sql, new { Id = id }, MapApplicationWithConfigs);
    }

    public async Task<List<Application>> GetAllAsync()
    {
        const string sql = "SELECT id, name, comments, created_at, updated_at FROM applications ORDER BY created_at DESC";
        
        return await _context.QueryAsync(sql, mapper: MapApplication);
    }

    public async Task<Application?> UpdateAsync(string id, ApplicationUpdate applicationData)
    {
        var now = DateTime.UtcNow;

        const string sql = @"
            UPDATE applications 
            SET name = @Name, comments = @Comments, updated_at = @UpdatedAt
            WHERE id = @Id
            RETURNING id, name, comments, created_at, updated_at";

        var parameters = new
        {
            Id = id,
            Name = applicationData.Name,
            Comments = applicationData.Comments,
            UpdatedAt = now
        };

        try
        {
            return await _context.QuerySingleOrDefaultAsync(sql, parameters, MapApplication);
        }
        catch (Exception ex) when (ex.Message.Contains("unique constraint") || ex.Message.Contains("duplicate"))
        {
            _logger.LogError(ex, "Unique constraint violation updating application");
            throw new InvalidOperationException($"Application with name '{applicationData.Name}' already exists", ex);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating application {Id}", id);
            throw;
        }
    }

    public async Task<bool> DeleteAsync(string id)
    {
        const string sql = "DELETE FROM applications WHERE id = @Id";
        
        var rowsAffected = await _context.ExecuteAsync(sql, new { Id = id });
        return rowsAffected > 0;
    }

    public async Task<int> DeleteMultipleAsync(List<string> ids)
    {
        if (!ids.Any())
        {
            return 0;
        }

        // Use a simpler approach with string concatenation for the IN clause
        // This is safe because we validate ULID format before calling this method
        var quotedIds = ids.Select(id => $"'{id}'");
        var sql = $"DELETE FROM applications WHERE id IN ({string.Join(", ", quotedIds)})";
        
        var rowsAffected = await _context.ExecuteAsync(sql);
        _logger.LogInformation("Deleted {Count} applications", rowsAffected);
        
        return rowsAffected;
    }

    public async Task<bool> ExistsAsync(string id)
    {
        const string sql = "SELECT COUNT(1) FROM applications WHERE id = @Id";
        
        var count = await _context.ExecuteScalarAsync<long>(sql, new { Id = id });
        return count > 0;
    }

    public async Task<bool> ExistsByNameAsync(string name, string? excludeId = null)
    {
        var sql = "SELECT COUNT(1) FROM applications WHERE name = @Name";
        object parameters = new { Name = name };

        if (!string.IsNullOrEmpty(excludeId))
        {
            sql += " AND id != @ExcludeId";
            parameters = new { Name = name, ExcludeId = excludeId };
        }

        var count = await _context.ExecuteScalarAsync<long>(sql, parameters);
        return count > 0;
    }

    private static Application MapApplication(IDataReader reader)
    {
        var createdAt = reader["created_at"];
        var updatedAt = reader["updated_at"];
        
        return new Application
        {
            Id = reader["id"].ToString()!,
            Name = reader["name"].ToString()!,
            Comments = reader["comments"] == DBNull.Value ? null : reader["comments"].ToString(),
            CreatedAt = createdAt == DBNull.Value ? DateTime.UtcNow : Convert.ToDateTime(createdAt),
            UpdatedAt = updatedAt == DBNull.Value ? DateTime.UtcNow : Convert.ToDateTime(updatedAt)
        };
    }

    private static ApplicationWithConfigs MapApplicationWithConfigs(IDataReader reader)
    {
        var createdAt = reader["created_at"];
        var updatedAt = reader["updated_at"];
        
        // Handle PostgreSQL array of configuration IDs
        var configurationIds = new List<string>();
        var configIdsValue = reader["configuration_ids"];
        if (configIdsValue != DBNull.Value)
        {
            var configIdsArray = (string[])configIdsValue;
            configurationIds = configIdsArray.ToList();
        }
        
        return new ApplicationWithConfigs
        {
            Id = reader["id"].ToString()!,
            Name = reader["name"].ToString()!,
            Comments = reader["comments"] == DBNull.Value ? null : reader["comments"].ToString(),
            CreatedAt = createdAt == DBNull.Value ? DateTime.UtcNow : Convert.ToDateTime(createdAt),
            UpdatedAt = updatedAt == DBNull.Value ? DateTime.UtcNow : Convert.ToDateTime(updatedAt),
            ConfigurationIds = configurationIds
        };
    }
}